### ISSUE TITLE
multi-node problem

### ISSUE THREAD

**Issue by sudahui** [stanford-futuredata/megablocks#18]
I have been using this MOE project and have achieved excellent results when running it on a single machine with 8 GPUs. However, I have noticed a significant drop in performance when using the same parameters on a multi-node setup(64 GPUs, 16 experts, tensor-parallel 4, same global batch size ). Additionally, I have encountered NCCL errors during the process.

I was wondering if you could provide some insights or optimizations for improving the performance of the MOE model in multi-node scenarios. Thanks

**Comment by tgale96** [comment]
Hi! Glad to hear you're using the project. 

You very likely have significantly less bandwidth between nodes (with Infiniband, presumably?) than you do within a node with NVLink. If you are going to use only expert sharding you'll want to have a large `ffn_hidden_size`. The ratio of computation to communication is proportional to this value, so having it be large helps to amortize the communication cost. If you're training a smaller model, you could try limiting the amount of expert sharding to reduce communication at the cost of increased weight storage per device. If the storage is a problem, I'd recommend using [ZeRO-2 or ZeRO-3](https://www.deepspeed.ai/tutorials/zero/). We have ZeRO-3 (aka, fully-sharded data parallelism) support, which you can turn on with the [moe_weight_parallelism](https://github.com/stanford-futuredata/megablocks/blob/main/megablocks/layers/arguments.py#L28) flag.

Basically, you're sharding for two reasons (1) storage and (2) arithmetic intensity. Any sharding (expert, tensor, weight) helps address (1). Expert model parallelism helps with (1) and (2). Sharding adds communication, so you want to shard only as much as is necessary.

What NCCL errors do you get? We run multi-node pretty reliably, but sometimes things like zombie processes can make NCCL error.

**Comment by sudahui** [comment]
Thank you for your suggestion!  I understand that inter-node communication is slower than intra-node, which requires adjusting the model configuration.

However, I noticed that even with the same hyperparameters(same global batch size, lr), the training loss increased significantly when moving from single-node to multi-node. My understanding is that while the throughput may decrease due to communication overhead, the loss should ideally remain comparable given the same batch size and model parameters.

So I wanted to check if you have any insights on what could be causing the higher loss in the multi-node setting with the same hyperparameters?  I want to ensure I am not missing something important. 

ps. NCCL problem in my environment has been solved which is caused by the PXN problem(close it works).

**Comment by tgale96** [comment]
Give [this](https://github.com/stanford-futuredata/megablocks/pull/20) branch a try. There was a small mathematical error with expert model parallelism that could cause deviations in loss for large world sizes.

Glad to hear your NCCL problem is resolved!

**Comment by tgale96** [comment]
Did the above change resolve your issue?

**Comment by tgale96** [comment]
I'm going to close this - feel free to re-open if you have more questions!
